src/client/Microsoft.Identity.Client/ManagedIdentity/ImdsV2ManagedIdentitySource.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Http.Retry;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class ImdsV2ManagedIdentitySource : AbstractManagedIdentity
+    {
+        private const string CsrMetadataPath = "/metadata/identity/getPlatformMetadata";
+
+        private static CsrMetadata _csrMetadata;
+
+        public static async Task<bool> GetCsrMetadataAsync(RequestContext requestContext)
```

Comment:
This method is doing too many things. I would recommend to break it down into validate, send request and process the response